### PR TITLE
Add ppe.onnx.grad to export Gradient Op

### PR DESCRIPTION
**Pull request by HiroakiMikami**

### CODE REVIEW

File: pytorch_pfn_extras/onnx/_grad.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        _grad_state.n_grad_call = None
+
+
+def grad(
+    output: torch.Tensor,
+    inputs: Tuple[torch.Tensor, ...],
+    retain_graph: Optional[bool] = None,
+    create_graph: bool = False,
+    allow_unused: bool = False,
+    only_inputs: bool = True,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/onnx/_grad.py:28]: Let's keep the same order as `torch.autograd.grad`.

```suggestion
    only_inputs: bool = True,
    allow_unused: bool = False,
```

https://pytorch.org/docs/stable/_modules/torch/autograd.html#grad

File: pytorch_pfn_extras/onnx/_grad.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                )
+        return _Gradient.apply(output, grad_output, *inputs_l)
+    else:
+        return torch.autograd.grad(
+            outputs=output,
+            inputs=inputs,
+            grad_outputs=grad_output,
+            retain_graph=retain_graph,
+            create_graph=create_graph,
+            allow_unused=allow_unused,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/onnx/_grad.py:97]: `only_inputs` is not passed.

```suggestion
            only_inputs=only_inputs,
            allow_unused=allow_unused,
```

File: pytorch_pfn_extras/onnx/_grad.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    inputs: Tuple[torch.Tensor, ...],
+    retain_graph: Optional[bool] = None,
+    create_graph: bool = False,
+    allow_unused: bool = False,
+    only_inputs: bool = True,
+) -> Tuple[Optional[torch.Tensor], ...]:
+    grad_output = torch.ones_like(output)
+
+    if torch.jit.is_tracing():
+        assert hasattr(_grad_state, "n_grad_call")
```
Comment (kmaehashi) [inline pytorch_pfn_extras/onnx/_grad.py:33]: Maybe better to raise an error with friendly message (e.g., `ppe.onnx.grad() can only be used in conjunction with export functions under `ppe.onnx`) instead of just an assertion?

File: pytorch_pfn_extras/onnx/_grad.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        inputs_l = list(inputs)
+        for i, input in enumerate(inputs):
+            input_name = f"Gradient_x_{i}_{n_grad_call}"
+            input_names.append(input_name)
+            inputs_l[i] = as_output(input_name, input)
+
+        class _Gradient(torch.autograd.Function):
+            @staticmethod
+            def forward(
+                ctx,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/onnx/_grad.py:52]:
```suggestion
                ctx: Any,
```

File: pytorch_pfn_extras/onnx/_grad.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                    inputs: List[torch.Tensor],
+                    grad_output: Optional[torch.Tensor],
+                    retain_graph: Optional[bool],
+                    create_graph: bool,
+                    allow_unused: bool,
+                ):
+                    return torch.autograd.grad(
+                        outputs=[output],
+                        inputs=inputs,
+                        grad_outputs=[grad_output],
```
Comment (kmaehashi) [inline pytorch_pfn_extras/onnx/_grad.py:69]: This should work? (align with the code in non-tracing case)

```suggestion
                        grad_outputs=grad_output,
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]
Thanks, LGTM except for minor comments!

File: pytorch_pfn_extras/onnx/_grad.py (hunk 5 above)
Comment (HiroakiMikami) [reply pytorch_pfn_extras/onnx/_grad.py:69]: no, torch script supports only the list case.

Comment (kmaehashi) [reply pytorch_pfn_extras/onnx/_grad.py:69]: I see, thanks for clarificaiton!

**Review by kmaehashi — APPROVED** [review]
LGTM!

**Comment by kmaehashi** [comment]
Could you check test failure? :pray:
```
E           Return value was annotated as having type List[Tensor] but is actually of type List[Optional[Tensor]]: |  
00:04:11.185542 STDOUT 2313] | E             File "/src/pytorch_pfn_extras/onnx/_grad.py", line 68 |  
00:04:11.185545 STDOUT 2313] | E                               allow_unused: bool, |  
00:04:11.185566 STDOUT 2313] | E                           ) -> List[torch.Tensor]: |  
00:04:11.185568 STDOUT 2313] | E                               return torch.autograd.grad( |  
00:04:11.185570 STDOUT 2313] | E                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~ |  
00:04:11.185589 STDOUT 2313] | E                                   outputs=[output], |  
00:04:11.185590 STDOUT 2313] | E                                   ~~~~~~~~~~~~~~~~~ |  
00:04:11.185592 STDOUT 2313] | E                                   inputs=inputs, |  
00:04:11.185605 STDOUT 2313] | E                                   ~~~~~~~~~~~~~~ |  
00:04:11.185607 STDOUT 2313] | E                                   grad_outputs=[grad_output],  # type: ignore |  
00:04:11.185648 STDOUT 2313] | E                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ |  
00:04:11.185650 STDOUT 2313] | E                                   retain_graph=retain_graph, |  
00:04:11.185654 STDOUT 2313] | E                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~ |  
00:04:11.185679 STDOUT 2313] | E                                   create_graph=create_graph, |  
00:04:11.185681 STDOUT 2313] | E                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~ |  
00:04:11.185683 STDOUT 2313] | E                                   allow_unused=allow_unused, |  
00:04:11.185702 STDOUT 2313] | E                                   ~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE |  
00:04:11.185703 STDOUT 2313] | E                               )
```

**Comment by HiroakiMikami** [comment]
I fixed the failure in my local environment.

**Comment by kmaehashi** [comment]
Thanks!
/test

File: pytorch_pfn_extras/onnx/_grad.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            ) -> Tuple[Optional[torch.Tensor], ...]:
+                @torch.jit.script
+                def _grad(
+                    output: torch.Tensor,
+                    inputs: List[torch.Tensor],
+                    grad_output: Optional[torch.Tensor],
+                    retain_graph: Optional[bool],
+                    create_graph: bool,
+                    allow_unused: bool,
+                ) -> List[Optional[torch.Tensor]]:
```
Comment (kmaehashi) [inline pytorch_pfn_extras/onnx/_grad.py:67]: Another mypy error :pray:
```
pytorch_pfn_extras/onnx/_grad.py:68: error: Incompatible return value type (got "Tuple[Tensor, ...]", expected "List[Optional[Tensor]]")  [return-value]
```

Comment (HiroakiMikami) [reply pytorch_pfn_extras/onnx/_grad.py:67]: hmm, the error occurs when using torch 1.8. and in torch 1.9,  using `Tuple[Tensor, ...]` causes mypy error.
I ignores the type hints in this tracing code path.

Comment (kmaehashi) [reply pytorch_pfn_extras/onnx/_grad.py]: Thanks, I agree ignoring type hints is ok here.
